### PR TITLE
HEEDLS-462 Reports - evaluation summary tables

### PR DESCRIPTION
**Pull request by ibrahimmunir14**
### JIRA link
[HEEDLS-462](https://softwiretech.atlassian.net/browse/HEEDLS-462)

### Description
Added evaluation summary section to Reports page, with cards containing evaluation summaries.
Each card displays the question and a summary list of the potential answers and the percentage of users that gave that answer.
If a question has no responses, a suitable message is displayed.
Takes into account the Reports filters (date range, job group, customisation, course category).

### Screenshots
**Laptop**

**Tablet**

**Mobile**

**Example of no responses for question:**
localhost_44363_TrackingSystem_Centre_Reports (3)

-----
### Developer checks
I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [x] Updated/added documentation in Swiki and/or Readme.
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

**Review by DanBloxham-sw — COMMENTED** [review]
Looks good, few minor points.

Also worth nothing the cards look a bit odd when they have a bunch of empty space from being stretched out in the card group (e.g. the Perform Faster? card next to the Time Saving Per Week card in the screen shot provided.) but I imagine all the spacing will get changed when the pie charts come in to it.

Could we also have some screen shots on smaller screen sizes?

**Review by DanBloxham-sw — APPROVED** [review]

File: DigitalLearningSolutions.Web.Tests/Helpers/EvaluationSummaryMappingHelperTests.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    using DigitalLearningSolutions.Data.Tests.TestHelpers;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.ViewModels.TrackingSystem.Centre.Reports;
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    public class EvaluationSummaryMappingHelperTests
+    {
+        [Test]
+        public void GenerateNumberWithLimitDisplayString_returns_expected_string_with_limit()
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/Helpers/EvaluationSummaryMappingHelperTests.cs:13]: I'm confused by this test name 👀

File: DigitalLearningSolutions.Web.Tests/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModelTests.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                0
+            );
+
+            // Then
+            model.Question.Should().Be("This question has never been seen!");
+            model.Data.Should().BeNull();
+        }
+
+        [Test]
+        public void EvaluationSummaryViewModel_sets_data_to_null_when_all_responses_are_noResponse()
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModelTests.cs:49]: We should show "No response" on the summary list - it's just a usual option like "Yes" or "No" from the admin's perspective

File: DigitalLearningSolutions.Web/Styles/trackingSystem/reports.scss [hunk 3]
```diff
@@ -93,3 +93,15 @@ $light-blue: #8EBAFF;
     transform: translate(-100%) rotate(-60deg);
   }
 }
+
+.eval-summary-row__key {
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Styles/trackingSystem/reports.scss:97]: Do we not want to use width defaults by NHS?

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/Reports/Index.cshtml [hunk 4]
```diff
@@ -39,6 +38,17 @@
     </div>
     <partial name="_ActivityTable.cshtml" model="@Model.UsageStatsTableViewModel" />
 
+    <hr />
+    <h2 class="nhsuk-heading-m">Evaluation summary</h2>
+
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/Reports/Index.cshtml:43]: This is extra, but I wonder if we want to add a paragraph here explaining what this section covers - worth asking HEE at Show & Tell tomorrow, then they can provide a copy if they think it'd be useful :)

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/Reports/_EvaluationSummaryCard.cshtml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+                @answer
+              </dt>
+              <dd class="nhsuk-summary-list__value eval-summary-row__value">
+                @percentage.ToString("0.0")%
+              </dd>
+            </div>
+          }
+        </dl>
+      } else {
+        <span>This question has not been answered yet</span>
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/Reports/_EvaluationSummaryCard.cshtml:25]: Full stop?

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        public EvaluationSummaryViewModel(string question, Dictionary<string, int> counts, int noResponseCount)
+        {
+            Question = question;
+            var totalResponses = counts.Sum(x => x.Value);
+            if (totalResponses > noResponseCount)
+            {
+                Data = counts
+                    .Select(x => new KeyValuePair<string, float>(x.Key, 100f * x.Value / totalResponses))
+                    .ToDictionary(x => x.Key, x => x.Value);
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs:16]: Any reason this has to be a dictionary in this case? Might be better as an object or a tuple list so we can improve naming & hence readability.. (and avoid going from dictionary to IEnumerable back to dictionary)🤔

**Review by stellake — CHANGES_REQUESTED** [review]

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/Reports/_EvaluationSummaryCard.cshtml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    <div class="nhsuk-card__description">
+      @if (Model.Data != null) {
+        <dl class="nhsuk-summary-list nhsuk-u-margin-bottom-3">
+          @foreach (var (answer, percentage) in Model.Data) {
+            <div class="nhsuk-summary-list__row">
+              <dt class="nhsuk-summary-list__key eval-summary-row__key">
+                @answer
+              </dt>
+              <dd class="nhsuk-summary-list__value eval-summary-row__value">
+                @percentage.ToString("0.0")%
```
[the remark this answers is not in the document]
Comment (ibrahimmunir14) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/Reports/_EvaluationSummaryCard.cshtml:19]: As this is just formatting do we need to test it?

File: DigitalLearningSolutions.Data/Models/TrackingSystem/EvaluationSummaryData.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿namespace DigitalLearningSolutions.Data.Models.TrackingSystem
+{
+    public class EvaluationSummaryData
+    {
+        public int Q1Yes { get; set; }
```
[the remark this answers is not in the document]
Comment (ibrahimmunir14) [reply DigitalLearningSolutions.Data/Models/TrackingSystem/EvaluationSummaryData.cs:5]: It seems like it's clear enough already but can change if you want.

File: DigitalLearningSolutions.Web/Styles/trackingSystem/reports.scss (hunk 3 above)
Comment (ibrahimmunir14) [reply DigitalLearningSolutions.Web/Styles/trackingSystem/reports.scss:97]: The default gives more width to the value and less to the key, but here we need more space for the key (question) and less for the value (percentage).

Comment (stellake) [reply DigitalLearningSolutions.Web/Styles/trackingSystem/reports.scss:97]: That makes sense :)

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+            if (model.ResponseCounts != null)
+            {
+                ResponsePercentages = model.ResponseCounts
+                    .Select(x => (question: x.response, percentage: 100f * x.count / model.TotalResponses));
+            }
+        }
+
+        public string Question { get; set; }
+
+        public IEnumerable<(string response, float percentage)>? ResponsePercentages { get; set; }
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs:21]: Do you think it'd be worth setting this as non-nullable (and default to empty array), then we don't need to handle empty array AND null value? I'm assuming both would be equivalent in front end.

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/ReportsViewModel.cs [hunk 10]
```diff
@@ -9,14 +9,21 @@
...
-        public ReportsViewModel(IEnumerable<PeriodOfActivity> activity, ReportsFilterModel filterModel)
+        public ReportsViewModel(
+            IEnumerable<PeriodOfActivity> activity,
+            ReportsFilterModel filterModel,
+            IEnumerable<EvaluationSummaryModel> evaluationSummaryModels
+        )
         {
             UsageStatsTableViewModel = new UsageStatsTableViewModel(activity);
             ReportsFilterModel = filterModel;
+            EvaluationSummaryViewModels =
```
Comment (stellake) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/ReportsViewModel.cs:20]: EvaluationSummaryBreakdown?

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs (hunk 9 above)
Comment (ibrahimmunir14) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs:21]: I feel like it's clearer if null represents the fact that there is no data?

Comment (stellake) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Centre/Reports/EvaluationSummaryViewModel.cs:21]: That's fair - can we add the handling the empty array case in frontend though?

**Review by stellake — APPROVED** [review]
Looks good - just need to add the extra paragraph as discussed. No need for rereview :)

**Pull request merged**